Add URLUtils class for URL resolution functionality

HamzaElzarw-2022 opened the pull request:
This PR introduces a URL utility class with a static resolve(URL base, String spec) method. It encapsulates logic previously implemented with deprecated URL(URL, String) constructors, improving clarity and future maintainability.

The utility is intended to be used for replacing deprecated URL constructors in PR #524.

closes #525

rzo1 commented on src/main/java/crawlercommons/utils/URLUtils.java:
```diff
@@ -0,0 +1,23 @@
+        try {
+            return base.toURI().resolve(spec).toURL();
+        } catch (Exception e) {
+            throw new MalformedURLException(e.getMessage());
```
I would wrap the underyling exception in the MUE, so the caller has an idea what exactly has gone wrong?

HamzaElzarw-2022 replied:
Thank you for the suggestion, I’ll update the code accordingly.

I also wanted to ask: would it be better to catch the specific exception types like URISyntaxException, IllegalArgumentException, and NullPointerException, or is it acceptable to catch a general Exception here since multiple types could be thrown?

HamzaElzarw-2022 replied:
I found that MalformedURLException doesn’t provide a constructor that takes an exception as an argument. So I used initCause() and casted the result back to MalformedURLException like this:

`throw (MalformedURLException) new MalformedURLException(e.getMessage()).initCause(e);`

Could you please confirm if this implementation is acceptable?

sebastian-nagel commented:
Hi @HamzaElzarw-2022, thanks and sorry for the delayed response.

I'd be fine with your implementation. It's a good starting point.

I've started to implement a unit test - for now it's a very early draft: #536. TODOs from the failing tests:
- address the bug with relative URLs (see #522)
- eventually allow the base URL to be null. This is allowed in the URL constructor if "spec" is an absolute URL:
  ```java
  new URL(null, "https://example.org/")
  ```
  Otherwise this should be documented. Also: throwing a NPE seems to be better in this case.
- there is an error if the "spec" starts with a colon:
  ```jshell
  jshell> new URL(new URL("https://example.org/foo/"), ":23")
  $7 ==> https://example.org/foo/:23

  jshell> new URI("https://example.org/foo/").resolve(":23")
  |  Exception java.lang.IllegalArgumentException: Expected scheme name at index 0: :23
  ...
  ```
- eventually implement a similar method for URIs as well
  
@HamzaElzarw-2022, let me know whether you want to continue to work on this PR. Otherwise, I'm going to merge it. We can address the open points later. And of course, feel free to pick from #536. Thanks!

HamzaElzarw-2022 commented:
Hello @sebastian-nagel, I will happily continue working on the PR.

HamzaElzarw-2022 commented:
Hello @sebastian-nagel, sorry for the very late follow-up.

I’ve addressed all the points you mentioned earlier, and all tests are now passing. Please review it when you have a moment.

To fix the bug with relative URLs described in NUTCH-797, I prepended the last segment to the spec to reverse the effect of URI.resolve().

HamzaElzarw-2022 commented on src/main/java/crawlercommons/utils/URLUtils.java:
```diff
@@ -0,0 +1,44 @@
+                }
+            }
+
+            return base.toURI().resolve(resolvedSpec).normalize().toURL();
```
I added `normalize()` in the first place to normalize in case it prepended `./` for when spec starts with a `:` but as you mentioned `URL::resolve` normalize relative paths any way. so in this case I think we should remove it as normalizing the base shouldn’t be part of this method’s responsibilities.

HamzaElzarw-2022 commented:
Great! After this is merged, I’ll open another PR to replace usages of URI.resolve() with this utility.

sebastian-nagel commented:
Thanks, @HamzaElzarw-2022 and @rzo1!
